reference-implementation/unittests/tosa.cpp

```diff
@@ -27,62 +28,192 @@ const float EPSILON = 5e-4;
+    Tensor2D<float, 1, 2> expected_result{-1.0, 3.0f};
+    Tensor2D<float, 1, 2> result = tosa::clamp(operand, min_value, max_value);
+
+    EXPECT_THAT(result, Pointwise(FloatEq(), expected_result));
+  }
+  {
+    Tensor3D<int64_t, 4, 2, 1> operand{-2, 2, -2, 3, 4, -5, 5, 5};
+    int64_t min_value = 1;
+    int64_t max_value = 3;
+    Tensor3D<int64_t, 4, 2, 1> expected_result{1, 2, 1, 3, 3, 1, 3, 3};
+    Tensor3D<int64_t, 4, 2, 1> result =
+        tosa::clamp(operand, min_value, max_value);
+
+    EXPECT_THAT(result, Pointwise(Eq(), expected_result));
+  }
+  {
+    Tensor4D<float, 1, 2, 3, 2> operand{-1.0, -0.5, 0.0, 2.0625, 2.875, 2.1875,
+                                        2.5,  -0.0, 2.0, 3.0,    -99.0, 99.0};
+    float min_value = 2.125;
+    float max_value = 2.875;
+    Tensor4D<float, 1, 2, 3, 2> expected_result{2.125, 2.125,  2.125, 2.125,
+                                                2.875, 2.1875, 2.5,   2.125,
+                                                2.125, 2.875,  2.125, 2.875};
+    Tensor4D<float, 1, 2, 3, 2> result =
+        tosa::clamp(operand, min_value, max_value);
+
+    EXPECT_THAT(result, Pointwise(FloatEq(), expected_result));
+  }
 }
 
 TEST(tosa, clz) {
-  Tensor<int32_t, 5> t0{0, 1, 0xC000, -1, -0x7FFFFFFF};
-  Tensor<int32_t, 5> expected{32, 31, 16, 0, 0};
-  Tensor<int32_t, 5> result = tosa::clz(t0);
-  EXPECT_THAT(result, Pointwise(Eq(), expected));
+  {
+    Tensor0D<int32_t> x{0};
+    Tensor0D<int32_t> expected_result{32};
+    Tensor0D<int32_t> result = tosa::clz(x);
+
+    EXPECT_THAT(result, Pointwise(Eq(), expected_result));
+  }
+  {
+    Tensor1D<int32_t, 2> x{1, -1};
+    Tensor1D<int32_t, 2> expected_result{31, 0};
+    Tensor1D<int32_t, 2> result = tosa::clz(x);
+
+    EXPECT_THAT(result, Pointwise(Eq(), expected_result));
+  }
+  {
+    Tensor2D<int32_t, 3, 2> x{-1328632289, 1459158945, -1912283137,
+                              627316066,   59808247,   42};
+    Tensor2D<int32_t, 3, 2> expected_result{0, 1, 0, 2, 6, 26};
+    Tensor2D<int32_t, 3, 2> result = tosa::clz(x);
+
+    EXPECT_THAT(result, Pointwise(Eq(), expected_result));
+  }
+  {
+    Tensor3D<int32_t, 2, 1, 2> x{0xC000, -0x7FFFFFFF, 132486298, -2104906602};
+    Tensor3D<int32_t, 2, 1, 2> expected_result{16, 0, 5, 0};
+    Tensor3D<int32_t, 2, 1, 2> result = tosa::clz(x);
+
+    EXPECT_THAT(result, Pointwise(Eq(), expected_result));
+  }
+  {
+    Tensor4D<int32_t, 1, 2, 2, 2> x{21845, 10922, 5461, 2730,
+                                    1365,  682,   341,  170};
+    Tensor4D<int32_t, 1, 2, 2, 2> expected_result{17, 18, 19, 20,
+                                                  21, 22, 23, 24};
+    Tensor4D<int32_t, 1, 2, 2, 2> result = tosa::clz(x);
+
+    EXPECT_THAT(result, Pointwise(Eq(), expected_result));
+  }
 }
 
 TEST(tosa, reciprocal) {
-  Tensor<float, 4> t0{1.0f, 2.0f, 3.0f, 4.0f};
-  Tensor<float, 4> expected{1.0f, 0.5f, 0.3333f, 0.25f};
+  {
+    Tensor0D<float> x{1.0f};
```

Comment:
Shouldn't negative values be tested too for this function?

Reply:
Yes!